### PR TITLE
[Fix] DLNA BROWSE issues.

### PR DESCRIPTION
**Pull request by BaronGreenback**
The Developer Tools for UPnP™ Technology's Device Validator reports up various errors when asked to the DLNA server. 
This PR contains the fixes to resolve these errors.

**Issue 1:**
Issuing a BROWSE browsemeta request against the DLNA server, when the item is the media folder of a user profile, returns a parent item and not 0 (it should be treated as a root item). I assume this code was written before user profiles.
**Fixed** This PR fixes this issue returning the correct value of 0.

**Issue 2:**
"Browse All": Browse("f137a2dd21bbc1b99aa5c0f6bf02a805", BROWSEDIRECTCHILDREN, "*", 0, 1, "") did not yield exactly one CDS-compliant media object in its result. Instantiated a total of 6 media objects.
**Fixed** The correct number of items are returned.

**Issue 3:**
"Browse All": Browse("continuewatching_f137a2dd21bbc1b99aa5c0f6bf02a805", BROWSEMETADATA, "*", 0, 0, "") returned a DIDL-Lite media object with parentID="0" when it should be "f137a2dd21bbc1b99aa5c0f6bf02a805".
**Fixed** Continue Watching's parent is now the top level folder.

### CODE REVIEW

**Comment by PrplHaz4** [comment]
Will be nice to have this fixed if not just for the good report card. I think browse used to be a little wonky with channel plugins.  I’ll try to test this here or on master in the next few days.

File: Emby.Dlna/ContentDirectory/ControlHandler.cs [hunk 1]
```diff
@@ -908,6 +908,16 @@ private QueryResult<ServerItem> GetMovieFolders(BaseItem item, User user, StubTy
                 }
             };
 
+            // If Limit is < 6 then we return too many items, so we must remove some.
```
Comment (cvium) [inline Emby.Dlna/ContentDirectory/ControlHandler.cs:911]: 6?

File: Emby.Dlna/ContentDirectory/ControlHandler.cs [hunk 2]
```diff
@@ -908,6 +908,16 @@ private QueryResult<ServerItem> GetMovieFolders(BaseItem item, User user, StubTy
...
+            if (limit != null && limit < array.Length)
```
Comment (cvium) [inline Emby.Dlna/ContentDirectory/ControlHandler.cs:912]:
```suggestion
            if (limit < array.Length)
```

File: Emby.Dlna/ContentDirectory/ControlHandler.cs [hunk 3]
```diff
@@ -908,6 +908,16 @@ private QueryResult<ServerItem> GetMovieFolders(BaseItem item, User user, StubTy
...
+            {
+                return new QueryResult<ServerItem>
+                {
+                    Items = array.ToArray().Take((int)limit).ToList(),
+                    TotalRecordCount = (int)limit
```
Comment (cvium) [inline Emby.Dlna/ContentDirectory/ControlHandler.cs:917]:
```suggestion
                    Items = array.ToArray().Take(limit.Value).ToList(),
                    TotalRecordCount = limit.Value
```

File: Emby.Dlna/ContentDirectory/ControlHandler.cs (hunk 1 above)
Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs:911]: There are 6 default items added to the list (see code prior). 
If limit is less than 6, six are still returned.

Will change comment.

File: Emby.Dlna/ContentDirectory/ControlHandler.cs (hunk 2 above)
Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs:912]: Well you learn something new every day!

Didn't know that the comparisons  > < <= and => with a null always return false. Will change code.

File: Emby.Dlna/ContentDirectory/ControlHandler.cs (hunk 3 above)
Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs:917]: In my defence - it was late at night.

**Comment by BaronGreenback** [comment]
> Will be nice to have this fixed if not just for the good report card. I think browse used to be a little wonky with channel plugins. I’ll try to test this here or on master in the next few days.

The remaining Device Validator batch test still fails as it's to do with ordering, and is expecting a specific subset of data.

File: Emby.Dlna/ContentDirectory/ControlHandler.cs [hunk 4]
```diff
@@ -908,6 +908,16 @@ private QueryResult<ServerItem> GetMovieFolders(BaseItem item, User user, StubTy
                 }
             };
 
+            // At this point, array contains six items. If Limit is < 6 then we return too many items, so the extra must be removed.
```
Comment (cvium) [inline Emby.Dlna/ContentDirectory/ControlHandler.cs:911]: I don't really get the stub type and why the limit should work this way. We're not really limiting the number of items, but the stub types?

File: Emby.Dlna/Didl/DidlBuilder.cs [hunk 5]
```diff
@@ -647,21 +652,26 @@ public void WriteFolderElement(XmlWriter writer, BaseItem folder, StubType? stub
...
-                if (context != null)
+                var parent = folder.DisplayParentId;
+                if (stubType == StubType.ContinueWatching)
+                {
+                    // Continue watching, parent is it's normal self.
+                    writer.WriteAttributeString("parentID", folder.Id.ToString("N", CultureInfo.InvariantCulture));
+                }
+                else if (parent.Equals(Guid.Empty)
+                    || folder.Parent.IsRoot
+                    || folder.Parent.DisplayParentId.Equals(Guid.Empty))
```
Comment (cvium) [inline Emby.Dlna/Didl/DidlBuilder.cs:663]:
```suggestion
                        || folder.Parent.IsRoot
                        || folder.Parent.DisplayParentId.Equals(Guid.Empty))
```

File: Emby.Dlna/ContentDirectory/ControlHandler.cs (hunk 4 above)
Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs:911]: I see what you mean. If that's the case, then the error lies with line #914 / 924 in that it should return 1 and not the array length.

Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs:911]: It looks to be slightly bigger issue than just this one. 

There are a number of routines that return this array structure for when the stub type does not match the expected content type. eg. GetMovieFolders has Latest, Movies, Collections, Favourites, and Genres.

It's found in three places.

Am looking at the specs.

Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs:911]: Failing on: Query: ObjectId 0 (Root), BrowseFlag = BrowseDirectChildren, RequestedCount = 0

Specification states (2.5.7.2 Arguments):-
```
TotalMatches: If BrowseFlag is set to “BrowseMetadata”, then TotalMatches MUST be set to 1.
Else if BrowseFlag is set to “BrowseDirectChildren”, then TotalMatches MUST be set to the total
number of objects in the object specified for the Browse() action (independent of the starting index
specified by the StartingIndex argument).
```
So if I'm reading that right, it means that the list of options shouldn't all get returned, and it should be trimmed the amount requested.

**Comment by BaronGreenback** [comment]
GitMerge failed - am rolling back.

File: Emby.Dlna/ContentDirectory/ControlHandler.cs (hunk 4 above)
Comment (cvium) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs]: I looked through the docs and our code. Wouldn't this change show an incorrect childCount for the root folder?

Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs]: Only if the RequestedCount is set, which i don't think it is (i'll double check).

This is more of a making the system meet the operational specs than making it work in Jellyfin,

Comment (BaronGreenback) [reply Emby.Dlna/ContentDirectory/ControlHandler.cs]: Have double checked - and it doesn't affect general browsing.

e.g, VLC sets a limit of 5000 by default.

**Comment by BaronGreenback** [comment]
Migrated into plugin

**Pull request closed without merging**